crates/shared/src/persistence/sqlite.rs

```diff
@@ -0,0 +1,211 @@
+        sqlx::query(
+            r#"
+            CREATE TABLE IF NOT EXISTS transactions (
+                id INTEGER PRIMARY KEY AUTOINCREMENT,
+                tx_data BLOB NOT NULL,
+                created_at DATETIME DEFAULT CURRENT_TIMESTAMP
+            );
+            "#,
+        )
+        .execute(&pool)
+        .await?;
+        Ok(Self { pool })
+    }
+
+    #[allow(dead_code)]
+    async fn clear(&self) -> Result<(), sqlx::Error> {
+        // Execute a SQL statement to delete all rows from the `transactions` table
+        sqlx::query("DELETE FROM transactions")
+            .execute(&self.pool)
+            .await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl BuilderPersistence for SqliteTxnDb {
+    async fn append(&self, tx_data: Vec<u8>) -> Result<(), sqlx::Error> {
+        sqlx::query(
+            r#"
+            INSERT INTO transactions (tx_data) VALUES (?);
+            "#,
+        )
+        .bind(tx_data)
+        .execute(&self.pool)
+        .await?;
+        Ok(())
+    }
+
+    async fn load(&self, timeout_after: Instant) -> Result<Vec<Vec<u8>>, sqlx::Error> {
+        // Convert Instant to SystemTime
+        let now = SystemTime::now();
+        let elapsed = timeout_after.elapsed();
+        let target_time = now - elapsed;
+
+        // Convert SystemTime to a format SQLite understands (RFC 3339)
+        let target_timestamp = DateTime::<Utc>::from(target_time)
+            .naive_utc()
+            .format("%Y-%m-%d %H:%M:%S")
+            .to_string();
+
+        let rows = sqlx::query(
+            r#"
+            SELECT id, tx_data, created_at FROM transactions
+            WHERE created_at <= ? ;
+            "#,
+        )
+        .bind(target_timestamp)
+        .fetch_all(&self.pool)
+        .await?;
+
+        let tx_data_list = rows
+            .into_iter()
+            .map(|row| row.get::<Vec<u8>, _>("tx_data"))
+            .collect();
+        Ok(tx_data_list)
+    }
+
+    async fn remove(&self, tx_data: Vec<u8>) -> Result<(), sqlx::Error> {
+        let result = sqlx::query(
+            r#"
+            DELETE FROM transactions WHERE tx_data = ?;
+            "#,
+        )
+        .bind(tx_data)
+        .execute(&self.pool)
+        .await?;
+
+        if result.rows_affected() > 0 {
+            Ok(())
+        } else {
+            Err(sqlx::Error::RowNotFound)
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::get_sqlite_test_db_path;
+    use super::BuilderPersistence;
+    use super::SqliteTxnDb;
+    use std::time::Instant;
+
+    /// This test checks we can set up sqlite properly
+    /// and can do basic append() and load()
+    #[tokio::test]
+    async fn test_persistence_append_and_load_txn() {
+        // Initialize the database
+        tracing::debug!(
+            "get_sqlite_test_db_path() = {:?}",
+            get_sqlite_test_db_path()
+        );
+        let db = SqliteTxnDb::new(get_sqlite_test_db_path()).await.expect(
```

Comment:
Does this create temporary files? If any assertion fails before the test completes, will those files be deleted?

Also, will it be better if we add an index or test name, etc. to distinguish the `get_sqlite_test_db_path` calls in tests and avoid race conditions (is it necessary?)? E.g., We can call `get_sqlite_test_db_path("test_append_and load")` in this function.

Reply:
It is updated to use the crate `tempfile` which will create a new temporary directory on the filesystem and the directory will be automatically deleted when it goes out of scope (if the returned TempDir value is dropped).